### PR TITLE
mvcc: tighten up watcher cancelation and revision handling

### PR DESCRIPTION
**Pull request by heyitsanthony**
Makes w.cur into w.minrev, the minimum revision for the next update, and
retries cancelation if the watcher isn't found (because it's being processed
by moveVictims).

Fixes: #5459

### CODE REVIEW

File: mvcc/watchable_store.go [hunk 1]
```diff
@@ -214,30 +214,41 @@ func (s *watchableStore) watch(key, end []byte, startRev int64, id WatchID, ch c
...
-		} else {
-			for _, wb := range s.victims {
-				if wb[wa] != nil {
-					slowWatcherGauge.Dec()
-					watcherGauge.Dec()
-					delete(wb, wa)
-					break
-				}
+			break
+		}
```
Comment (gyuho) [inline mvcc/watchable_store.go]: `break` here?

Comment (gyuho) [inline mvcc/watchable_store.go]: nvm.

**Comment by gyuho** [comment]
test failure is related to this change?

**Comment by heyitsanthony** [comment]
@gyuho not sure; the test that's failing (TestGRPCStreamRequireLeader) is already broken in master but the failure mode is "etcdserver: not capable" and not a timeout.

**Comment by gyuho** [comment]
Oh you already created an issue. Got it.

File: mvcc/watchable_store.go [hunk 2]
```diff
@@ -307,7 +318,7 @@ func (s *watchableStore) moveVictims() (moved int) {
 		// try to send responses again
 		for w, eb := range wb {
 			select {
-			case w.ch <- WatchResponse{WatchID: w.id, Events: eb.evs, Revision: w.cur}:
+			case w.ch <- WatchResponse{WatchID: w.id, Events: eb.evs, Revision: w.minRev - 1}:
```
Comment (gyuho) [inline mvcc/watchable_store.go]: can we have some comments on why `-1`?

**Comment by xiang90** [comment]
lgtm

**Comment by heyitsanthony** [comment]
still broken following the api capability fixes, will investigate

**Comment by heyitsanthony** [comment]
It was deadlocking waiting for the recvLoop to finish since the stream recv doesn't close until the handler exits; now it only waits on the send loop. The watcher wg patch is still necessary since the watchStream close needs to be synchronous to avoid spurious goroutine leak errors and must run concurrently with the recvloop goroutine to avoid deadlock

**Pull request merged**